### PR TITLE
Movement unit tests & refactor

### PR DESCRIPTION
**Pull request by PabloMansanet**
This started as unit tests for the movement module, but I've found several small issues that I've fixed along the way. To name a couple off the top of my head:

* Multiple word motions weren't working, and it was difficult to solve elegantly without changing the API so it takes an initial Range as opposed to an index.
* Some edge cases involving newlines weren't properly handled.
* Failure to move used to reset the range, instead of leaving it untouched like kak/vim.

I've also refactored a few functions to be more functional and have less difficult to track mutation going on.

The current draft only contains unit tests (and thus  correct behaviour to the best of my knowledge) for the forward word move, but the backward one is still WIP.

### CODE REVIEW

File: helix-core/src/movement.rs [hunk 1]
```diff
@@ -58,80 +104,55 @@ pub fn move_vertically(
...
-        }
-
-        let mut ch = slice.char(begin);
-        let next = slice.char(begin + 1);
-
-        // if we're at the end of a word, or on whitespce right before new one
-        if categorize(ch) != categorize(next) {
-            begin += 1;
-        }
+pub fn move_next_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
```
Comment (PabloMansanet) [inline helix-core/src/movement.rs:117]: This is the only of the "word movement" functions that is finished; the other two would be modeled after this.

File: helix-core/src/movement.rs [hunk 2]
```diff
@@ -1,18 +1,60 @@
...
+    fn outside(&self, text: RopeSlice) -> bool;
+    /// The next character after this belongs
+    /// to a different `Category`
+    fn is_category_boundary(&self, text: RopeSlice) -> bool;
+    fn is(&self, category: Category, text: RopeSlice) -> bool;
+    fn category(&self, text: RopeSlice) -> Option<Category>;
+}
+
+impl SliceHelpers for usize {
+    fn outside(&self, text: RopeSlice) -> bool {
```
Comment (kirawi) [inline helix-core/src/movement.rs:31]: Could these methods be `inline`d?

File: helix-core/src/movement.rs [hunk 3]
```diff
@@ -544,87 +542,129 @@ mod test {
...
+                    }
+                };
+                assert_eq!(range, expected_end, "Case failed: [{}]", sample);
+            }
+        }
+    }
+
+    #[test]
+    fn test_behaviour_when_moving_to_start_of_previous_words() {
+        let tests = array::IntoIter::new([
```
Comment (pickfire) [inline helix-core/src/movement.rs:641]: I think we can just use `vec!` since it's tests and perf does not matter.

File: helix-core/src/movement.rs [hunk 4]
```diff
@@ -544,87 +542,129 @@ mod test {
...
+            ("Basic backward motion from the middle of a word",
+                vec![(Motion::PrevStart(1), Range::new(3, 3), Range::new(3, 0))]),
```
Comment (pickfire) [inline helix-core/src/movement.rs:643]: Why is motion always `PrevStart`?

File: helix-core/src/movement.rs [hunk 5]
```diff
@@ -544,87 +542,129 @@ mod test {
...
-            TestCase("\n   \n   \n Jumping through alternated space blocks and newlines spans the space blocks",
+            ("\n   \n   \n Jumping through alternated space blocks and newlines selects the space blocks",
                 vec![
                     (Motion::NextStart(1), Range::new(0, 0), Range::new(1, 3)),
                     (Motion::NextStart(1), Range::new(1, 3), Range::new(5, 7)),
                 ]),
-            TestCase("ヒーリクス multibyte characters behave as normal characters",
+            ("ヒーリクス multibyte characters behave as normal characters",
                 vec![
                     (Motion::NextStart(1), Range::new(0, 0), Range::new(0, 5)),
```
Comment (pickfire) [inline helix-core/src/movement.rs:619]: Same for this, why is it always `NextStart`?

**Comment by kirawi** [comment]
I think `w` is incorrect, as it moves to ~end of the word (there's room for a fix there to correct it) rather than the start, which should be `e`'s job, no?

File: helix-core/src/movement.rs (hunk 2 above)
Comment (PabloMansanet) [reply helix-core/src/movement.rs:31]: I usually don't mark private functions as `inline` since you're not giving the compiler that much to work with, since there's no decision to be made with regards to inlining across crate boundaries. If they end up becoming public and we're not using LTO though, why not.

**Comment by kirawi** [comment]
Ah, the docs weren't that clear. I thought the cursor was supposed to go to the first character of each word.

File: helix-core/src/movement.rs [hunk 6]
```diff
@@ -58,80 +104,54 @@ pub fn move_vertically(
...
-    for _ in 0..count {
-        if begin + 1 == slice.len_chars() {
-            return None;
-        }
-
-        let mut ch = slice.char(begin);
-        let next = slice.char(begin + 1);
+pub fn move_next_word_start(slice: RopeSlice, range: Range, count: usize) -> Range {
+    let last_index = slice.len_chars().saturating_sub(1);
+    let grouped_categories = [Category::Punctuation, Category::Word, Category::Whitespace];
```
Comment (kirawi) [inline helix-core/src/movement.rs:119]: Why is this necessary? If it is necessary, it would probably better to use a match instead.

**Comment by kirawi** [comment]
Additionally, could you add documentation explaining what is the purpose of each function, and elaborate on how it generally works for future developers? I don't quite understand it myself yet, and it's been quite difficult to create a mental model of how everything fits together because I'm not completely sure what each function is for or why it was designed a certain way (i.e. why do we add `+ 1` to `range.head` and why do we do multiple `skip_while`s).

File: helix-core/src/movement.rs [hunk 7]
```diff
@@ -29,15 +71,19 @@ pub fn move_horizontally(
             nth_next_grapheme_boundary(text, pos, count).min(end)
         }
     };
-    Range::new(if extend { range.anchor } else { pos }, pos)
```
Comment (archseer) [inline helix-core/src/movement.rs:32]: This is a common pattern, I'm planning to extract it into something like `range.transform(cond, pos)` I think.

```rust
range.transform(behaviour == SelectionBehaviour::Extend, pos)
```

File: helix-core/src/movement.rs (hunk 3 above)
Comment (archseer) [reply helix-core/src/movement.rs:641]: I think IntoIter is fine for now since we'll be able to use [plain arrays without the wrapper](https://twitter.com/m_ou_se/status/1385966446254166020) when 1.53 is stable

File: helix-term/src/commands.rs [hunk 8]
```diff
@@ -131,7 +133,7 @@ pub fn move_char_left(cx: &mut Context) {
             range,
             Direction::Backward,
             count,
-            false, /* extend */
+            movement::SelectionBehaviour::Displace,
```
Comment (archseer) [inline helix-term/src/commands.rs:136]: What if we call it `Movement::Extend`/`Move`?

**Comment by PabloMansanet** [comment]
Thanks for the comments! Closing it for now while I address the review, I'll put it up again with @archseer 's proposed changes and the rest of the tests + more refactor.
